internal/venice-client-common/src/test/java/com/linkedin/venice/stats/metrics/MetricEntityStateGenericTest.java

```diff
@@ -0,0 +1,216 @@
+    doReturn(dimensionsSet).when(mockMetricEntity).getDimensionsList();
+    mockSensor = mock(Sensor.class);
+    baseDimensionsMap = new HashMap<>();
+    baseDimensionsMap.put(VENICE_REQUEST_METHOD, MULTI_GET_STREAMING.getDimensionValue());
+
+    // build a map of inputDimensions
+    testInputDimensions = new HashMap<>();
+    testInputDimensions.put(VENICE_STORE_NAME, "store1");
+    testInputDimensions.put(VENICE_CLUSTER_NAME, "cluster1");
+
+    AttributesBuilder attributesBuilder = Attributes.builder();
+    for (Map.Entry<VeniceMetricsDimensions, String> entry: testInputDimensions.entrySet()) {
+      attributesBuilder.put(mockOtelRepository.getDimensionName(entry.getKey()), entry.getValue()).build();
+    }
+    for (Map.Entry<VeniceMetricsDimensions, String> entry: baseDimensionsMap.entrySet()) {
+      attributesBuilder.put(mockOtelRepository.getDimensionName(entry.getKey()), entry.getValue()).build();
+    }
+    expectedAttributesForTestInputDimensions = attributesBuilder.build();
+  }
+
+  @Test
+  public void testRecordOtelMetricCounter() {
+    LongCounter longCounter = mock(LongCounter.class);
+    when(mockMetricEntity.getMetricType()).thenReturn(MetricType.COUNTER);
+
+    MetricEntityState metricEntityState =
+        new MetricEntityStateGeneric(mockMetricEntity, mockOtelRepository, baseDimensionsMap);
+    metricEntityState.setOtelMetric(longCounter);
+
+    Attributes attributes = Attributes.builder().put("key", "value").build();
+    metricEntityState.recordOtelMetric(10, attributes);
+
+    verify(longCounter, times(1)).add(10, attributes);
+  }
+
+  @Test
+  public void testRecordMetricsWithBothOtelAndTehuti() {
+    DoubleHistogram doubleHistogram = mock(DoubleHistogram.class);
+    when(mockMetricEntity.getMetricType()).thenReturn(HISTOGRAM);
+
+    MetricEntityStateGeneric metricEntityState =
+        new MetricEntityStateGeneric(mockMetricEntity, mockOtelRepository, baseDimensionsMap);
+    metricEntityState.setOtelMetric(doubleHistogram);
+    metricEntityState.setTehutiSensor(mockSensor);
+
+    // make a copy of the testInputDimensions as it will be modified in the test
+    Map<VeniceMetricsDimensions, String> testInputDimensions = new HashMap<>(this.testInputDimensions);
+
+    // called 0 times
+    verify(doubleHistogram, times(0)).record(20.0, expectedAttributesForTestInputDimensions);
+    verify(mockSensor, times(0)).record(20.0);
+
+    // called 1 time
+    metricEntityState.record(20.0, testInputDimensions);
+    verify(doubleHistogram, times(1)).record(20.0, expectedAttributesForTestInputDimensions);
+    verify(mockSensor, times(1)).record(20.0);
+
+    // called 2 times
+    metricEntityState.record(20.0, testInputDimensions);
+    verify(doubleHistogram, times(2)).record(20.0, expectedAttributesForTestInputDimensions);
+    verify(mockSensor, times(2)).record(20.0);
+
+    // test without full dimensions
+    testInputDimensions.remove(VENICE_CLUSTER_NAME);
```

Comment:
Don't we throw `IllegalArgumentException` if we remove a dimension? The value `20.0` can't be recorded in this case, is that right? e.g. In your later test, `testGetAttributes`, case 2, seems to be a similar scenario and the code expects a `IllegalArgumentException` to be thrown

Reply:
Discussing it at https://github.com/linkedin/venice/pull/1667/files/c61e0de8c71d07aaaa720bb2eec1cd68b9319955#r2047146002